### PR TITLE
Add IMDS fallback with DescribeInstances.

### PR DESCRIPTION
**Pull request by jefchien**
# Description of the issue
Related to https://github.com/aws/amazon-cloudwatch-agent/issues/1101.

Instance Metadata Service (IMDS) provides key information about the local EC2 instance including region and instance ID. The CloudWatch Agent relies on IMDS to retrieve those fields, which it uses as metric dimensions. As a security best practice, EKS [recommends](https://aws.github.io/aws-eks-best-practices/security/docs/iam/#restrict-access-to-the-instance-profile-assigned-to-the-worker-node) that the [hop limit](https://en.wikipedia.org/wiki/Hop_(networking)) on worker nodes is set to 1. This effectively disables IMDS within containers as any response for a call made within a container will require at least 2 hops.

In the cases where IMDS is not available or disabled, the agent needs an additional fallback mechanism to retrieve the metadata without IMDS. For EC2 instances, the default AWS hostname follows a [fixed format](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-naming.html) and will be prefixed with either the private IPv4 address (`ip-10-24-34-0.ec2.internal`) or the instance ID (`i-0123456789abcdef.us-west-2.compute.internal`) based on the hostname type configured on the instance. In these cases, `DescribeInstances` can be used to get the instance metadata instead.

# Description of changes
Adds a fallback metadata provider that parses the hostname to determine an EC2 filter and makes a call to `DescribeInstances` with that filter. Consolidates the metadata provider functions into a single package.

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
Added unit tests and built/installed the agent on an EC2 host with IMDS disabled. Configured `common-config.toml` to point to shared credential file.

`fmt.Print` to stdout isn't being logged for some reason. The IMDS logging is from the AWS SDK.

Without DescribeInstances permissions
```
> sudo /opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent-ctl -a fetch-config -c file:test.json -s
****** processing amazon-cloudwatch-agent ******
2024/04/17 15:35:28 I! WARN: failed to get session token, falling back to IMDSv1: 403 Forbidden: Forbidden
	status code: 403, request id: 
caused by: EC2MetadataError: failed to make EC2Metadata request
<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
 <head>
  <title>403 - Forbidden</title>
 </head>
 <body>
  <h1>403 - Forbidden</h1>
 </body>
</html>

	status code: 403, request id: 
2024/04/17 15:35:29 E! Please check if you can access the metadata service. For example, on linux, run 'wget -q -O - http://169.254.169.254/latest/meta-data/instance-id && echo' 
```

With DescribeInstances permissions
```
> sudo /opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent-ctl -a fetch-config -c file:test.json -s
****** processing amazon-cloudwatch-agent ******
2024/04/17 16:23:46 I! WARN: failed to get session token, falling back to IMDSv1: 403 Forbidden: Forbidden
	status code: 403, request id: 
caused by: EC2MetadataError: failed to make EC2Metadata request
<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
 <head>
  <title>403 - Forbidden</title>
 </head>
 <body>
  <h1>403 - Forbidden</h1>
 </body>
</html>

	status code: 403, request id: 
Got Home directory: /root I! Set home dir Linux: /root I! SDKRegionWithCredsMap region: I! Trying to detect region from ec2 D! [EC2] Found active network interface I! imds retry client will retry 1 times D! should retry true for imds error : EC2MetadataError: failed to make EC2Metadata request <?xml version="1.0" encoding="iso-8859-1"?> <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"> <html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en"> <head> <title>403 - Forbidden</title> </head> <body> <h1>403 - Forbidden</h1> </body> </html> status code: 403, request id: D! should retry true for imds error : EC2MetadataError: failed to make EC2Metadata request <?xml version="1.0" encoding="iso-8859-1"?> <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"> <html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en"> <head> <title>403 - Forbidden</title> </head> <body> <h1>403 - Forbidden</h1> </body> </html> status code: 403, request id: D! should retry true for imds error : EC2MetadataError: failed to make EC2Metadata request <?xml version="1.0" encoding="iso-8859-1"?> <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"> <html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en"> <head> <title>403 - Forbidden</title> </head> <body> <h1>403 - Forbidden</h1> </body> </html> status code: 403, request id: D! should retry true for imds error : EC2MetadataError: failed to make EC2Metadata request <?xml version="1.0" encoding="iso-8859-1"?> <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"> <html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en"> <head> <title>403 - Forbidden</title> </head> <body> <h1>403 - Forbidden</h1> </body> </html> status code: 403, request id: Successfully fetched the config and saved in /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.d/file_test2.json.tmp
Start configuration validation...
2024/04/17 16:23:46 Reading json config file path: /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.d/file_test2.json.tmp ...
2024/04/17 16:23:46 I! Valid Json input schema.
2024/04/17 16:23:47 I! WARN: failed to get session token, falling back to IMDSv1: 403 Forbidden: Forbidden
	status code: 403, request id: 
caused by: EC2MetadataError: failed to make EC2Metadata request
<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
 <head>
  <title>403 - Forbidden</title>
 </head>
 <body>
  <h1>403 - Forbidden</h1>
 </body>
</html>

	status code: 403, request id: 
2024/04/17 16:23:47 D! ec2tagger processor required because append_dimensions is set
2024/04/17 16:23:47 D! delta processor required because metrics with diskio or net are set
2024/04/17 16:23:47 D! ec2tagger processor required because append_dimensions is set
2024/04/17 16:23:47 Configuration validation first phase succeeded
I! Detecting run_as_user...
D! [EC2] Found active network interface
I! imds retry client will retry 1 times
D! should retry true for imds error : EC2MetadataError: failed to make EC2Metadata request
<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
 <head>
  <title>403 - Forbidden</title>
 </head>
 <body>
  <h1>403 - Forbidden</h1>
 </body>
</html>

	status code: 403, request id: 
D! should retry true for imds error : EC2MetadataError: failed to make EC2Metadata request
<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
 <head>
  <title>403 - Forbidden</title>
 </head>
 <body>
  <h1>403 - Forbidden</h1>
 </body>
</html>

	status code: 403, request id: 
D! should retry true for imds error : EC2MetadataError: failed to make EC2Metadata request
<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
 <head>
  <title>403 - Forbidden</title>
 </head>
 <body>
  <h1>403 - Forbidden</h1>
 </body>
</html>

	status code: 403, request id: 
D! should retry true for imds error : EC2MetadataError: failed to make EC2Metadata request
<?xml version="1.0" encoding="iso-8859-1"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
	"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
 <head>
  <title>403 - Forbidden</title>
 </head>
 <body>
  <h1>403 - Forbidden</h1>
 </body>
</html>

	status code: 403, request id: 
/opt/aws/amazon-cloudwatch-agent/bin/amazon-cloudwatch-agent -schematest -config /opt/aws/amazon-cloudwatch-agent/etc/amazon-cloudwatch-agent.toml
Configuration validation second phase succeeded
Configuration validation succeeded
```

# Requirements
_Before commit the code, please do the following steps._
1. Run `make fmt` and `make fmt-sh`
2. Run `make lint`

### CODE REVIEW

File: internal/metadata/ec2/chain.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	for _, provider := range p.providers {
+		providerIDs = append(providerIDs, provider.ID())
+	}
+	return fmt.Sprintf("Chain [%s]", strings.Join(providerIDs, ","))
+}
+
+func (p *chainMetadataProvider) Get(ctx context.Context) (metadata *Metadata, err error) {
+	for _, provider := range p.providers {
+		if metadata, err = provider.Get(ctx); err == nil {
+			log.Printf("D! [MetadataProvider/%s] got metadata: %#v\n", provider.ID(), metadata)
```
[the remark this answers is not in the document]
Comment (jefchien) [reply internal/metadata/ec2/chain.go]: Removed all logging from providers. The callers can determine how they want to log the errors.

**Review by sethAmazon — APPROVED** [review]

**Review by okankoAMZ — APPROVED** [review]
Looks good! Left few questions

File: plugins/processors/ec2tagger/ec2tagger.go [hunk 2]
```diff
@@ -413,7 +425,7 @@ func (t *Tagger) setStarted() {
 
 /*
 Retrieve metadata from IMDS and use these metadata to:
-* Extract InstanceID, ImageID, InstanceType to create custom dimension for collected metrics
+* Extract InstanceID, imageID, InstanceType to create custom dimension for collected metrics
```
[the remark this answers is not in the document]
Comment (jefchien) [reply plugins/processors/ec2tagger/ec2tagger.go:428]: oops

File: internal/metadata/ec2/non_imds.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+	filter, err := filterFromHostnamePrefix(prefix)
+	if err != nil {
+		return nil, "", err
+	}
+	return filter, region, nil
+}
+
+func fromReservation(reservation ec2.Reservation) (*Metadata, error) {
+	instanceCount := len(reservation.Instances)
+	if instanceCount == 0 || instanceCount > 1 {
```
[the remark this answers is not in the document]
Comment (jefchien) [reply internal/metadata/ec2/non_imds.go:119]: Yes. If we find multiple instances, then our filter is wrong.

**Pull request merged**